power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -714,50 +766,113 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
   public:
     template <class Component> using UpdateType = typename Component::UpdateType;
 
-    static bool is_update_independent(ConstDataset const& update_data) {
-        // If the batch size is (0 or) 1, then the update data for this component is 'independent'
-        if (update_data.batch_size() <= 1) {
-            return true;
-        }
+    static UpdateCompIndependence check_components_independence(ConstDataset const& update_data) {
+        auto check_ids_na = [](auto const& all_spans) {
+            std::vector<std::vector<bool>> ids_na{};
+            for (const auto& span : all_spans) {
+                for (const auto& obj : span) {
+                    std::vector<bool> id_na{};
+                    if constexpr (requires { obj.id; }) {
+                        id_na.push_back(obj.id == na_Idx);
+                    }
+                    ids_na.push_back(id_na);
+                }
+            }
+            return ids_na;
+        };
 
-        auto const process_buffer_span = []<typename CT>(auto const& all_spans) -> bool {
+        auto process_buffer_span = [check_ids_na]<typename CT>(auto const& all_spans, UpdateCompProperties& result) {
             // Remember the first batch size, then loop over the remaining batches and check if they are of the same
             // length
-            auto const elements_per_scenario = static_cast<Idx>(all_spans.front().size());
-            bool const uniform_batch = std::ranges::all_of(all_spans, [elements_per_scenario](auto const& span) {
-                return static_cast<Idx>(span.size()) == elements_per_scenario;
+            std::vector<std::vector<bool>> const ids_na = check_ids_na(all_spans);
+            result.has_id = !std::ranges::all_of(ids_na, [](const std::vector<bool>& vec) { return vec.empty(); });
+            result.ids_all_na = std::ranges::all_of(ids_na, [](const std::vector<bool>& vec) {
+                return std::ranges::all_of(vec, [](bool const& obj) { return obj == true; });
             });
-            if (!uniform_batch) {
-                return false;
-            }
-            if (elements_per_scenario == 0) {
-                return true;
-            }
+            result.ids_part_na = std::ranges::any_of(ids_na, [](const std::vector<bool>& vec) {
+                return std::ranges::any_of(vec, [](bool const& obj) { return obj == true; }) &&
+                       std::ranges::any_of(vec, [](bool const& obj) { return obj == false; });
```

Comment:
```suggestion
                return std::ranges::any_of(vec, [](bool const& obj) { return obj; }) &&
                       std::ranges::any_of(vec, [](bool const& obj) { return !obj; });
```
or even by accumulating `obj ? 1 : 0` and checking that the resultant is less than to size of the vector.

you can also use something along the lines of
```suggestion
                auto comp = [](bool const& obj) { return obj; };
                return std::ranges::any_of(vec, comp) && !std::ranges::all_of(vec, comp);
```

Reply:
Accumulating and `all_of` has higher complexity than `any_of` expectedly.

Reply:
why? https://en.cppreference.com/w/cpp/algorithm/all_any_none_of `any_of` just early-returns of `if(pred())`, `all_of` early-returns on `if(!pred())`

Reply:
Then it makes no difference which one to use, no?

Reply:
indeed except for readability

Reply:
I guess that's a matter of opinion.